chainer/functions/math/basic_math.py

```diff
@@ -466,22 +511,65 @@ def check_type_forward(self, in_types):
         type_check.expect(in_types[0].dtype.kind == 'f')
 
     def forward(self, x):
-        value = _preprocess_const(x[0], self.value)
-        return utils.force_array(x[0] ** value, x[0].dtype),
+        self.retain_inputs((0,))
 
-    def backward_cpu(self, x, gy):
-        val_1 = _preprocess_const(x[0], self.value - 1)
-        gx = utils.force_type(x[0].dtype, self.value) * (x[0] ** val_1) * gy[0]
-        return utils.force_array(gx),
+        y = x[0] ** _preprocess_const(x[0], self.value)
 
-    def backward_gpu(self, x, gy):
-        value = _preprocess_const(x[0], self.value)
-        gx = cuda.elementwise(
-            'T x, T gy, T value', 'T gx',
-            'gx = value * pow(x, value - 1) * gy',
-            'pow_var_const_bwd')(x[0], gy[0], value)
+        if isinstance(x[0], numpy.ndarray):
+            y = utils.force_array(y, x[0].dtype)
+        return y,
+
+    def backward(self, indexes, gy):
+        inputs, = self.get_retained_inputs()
+        return PowVarConstGrad(self.value).apply((inputs[0], gy[0]))
+
+
+class PowVarConstGrad(function_node.FunctionNode):
+
+    def __init__(self, value):
+        self.value = value
+        self.val = self.val_1 = None
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        type_check.expect(
+            in_types[0].dtype.kind == 'f',
+            in_types[0].dtype == in_types[1].dtype,
+            in_types[0].shape == in_types[1].shape
+        )
+
+    def forward(self, inputs):
```

Comment:
Since logic overlap (unpacking and retaining) and the the remaining code is relatively thin, I thought this implementation was easier to read. Do you want me to change it?

Reply:
Yes, because it would be generally better to follow the framework's system and what its designer intends. For example, when some facilities might be added to `forward_cpu` and `forward_gpu` in the future, the current implementation could not follow the possible change.